### DISCUSSION TITLE
Meson or CMake based build system

### DISCUSSION THREAD

**Discussion by wbehrens-on-gh** [ImageMagick/ImageMagick#5102] [Development]
### Is your feature request related to a problem? Please describe.

I'll be blunt, the current build system for imagemagick kinda sucks, at least from an end user perspective. I've been trying to fix the conan.io package for imagemagick for about a week now as it doesn't officially support windows and I've realized some stuff.

1. Autotools, while flexible, is clearly too much to be documented effectively as finding any documentation on the build system for imagemagick that isn't outdated is a challenge. Plus I think I goes without saying that autotools by todays standards is overly complicated.

2. Windows port? why is it it's own library/repo? why does it require separate build scripts? why is there no easy way to pin it at a specific version?

### Describe the solution you'd like

I'd be willing to try and implement a new cmake/meson build system to try and modernize this a bit, I know some people have tried in the past but their efforts appeared to have dried up. I would have to know though if the imagemagick team would be willing to change the build system though if someone did submit a patch for it? and what exactly would this build system need to do in order for it to be merged?

### Describe alternatives you've considered

Another thing I've thought about is possibly documenting the windows port better to at least tell people how to make use of the existing autotools system with cygwin to cross compile to visual studio.

**Comment by wbehrens-on-gh** [comment]
And I would also be wondering if I could submit these patches to imagemagick6?

**Comment by urban-warrior** [comment]
The developers are of course interested in a modern cmake build.  Create the requisite files and post a link where we can download and test.  Once proven they are a suitable replacement of the existing automake/autconf build, we would likely adopt the cmake build.  Be sure to export your cmake build under the ImageMagick license.  Thanks.

**Comment by vincent-nv** [comment]
It can be really interesting to create portfile to in vcpkg !! Idealy with CMake buildsystem.

**Comment by Cyriuz** [comment]
I adapted the work by @MarcoMartins86 in #743 to work with the latest version of ImageMagick and added some initial install support which I intend to use for making the conan package work on all platforms. Initially I'll probably just make it a patch in the conan recipe but I'll leave a link to it here if anyone else wants to help out since there is quite a lot left to do (tests, missing some dependencies, a lot of options and documentation etc)

I understand that the developers want to see a complete replacement but I think the amount of stuff that has amassed in the old build scrips over the years will take a while to port over, especially if the contributor is on another platform/doesn't use the features. Maybe you could consider taking over a solution that covers at least the basics? That way it is easier to find for other people that want to contribute.

Anyway here's the fork I'll be trying to polish a bit more: https://github.com/Cyriuz/ImageMagick

**Comment by urban-warrior** [comment]
We're interested in `cmake` support for ImageMagick but adopting will need to be staged.  First, let the community contribute to this repo until we have robust scripts that's "98% there".   For example, we cloned this repo on a Fedora system and the build failed with 

```
/usr/include/CL/cl_version.h:22:9: note: ‘#pragma message: cl_version.h: CL_TARGET_OPENCL_VERSION is not defined. Defaulting to 300 (OpenCL 3.0)’
...
home/cristy/Downloads/ImageMagick-main/MagickCore/cache.c: In function ‘Acquire
PixelCacheNexus’:
../ImageMagick-main/MagickCore/memory-private.h:30:33: error
: ‘MAGICKCORE_SIZEOF_VOID_P’ undeclared (first use in this function); did you me
an ‘MAGICKCORE_SIZEOF_OFF_T’?
   30 | #  define CACHE_LINE_SIZE  (8 * MAGICKCORE_SIZEOF_VOID_P)
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~
.../ImageMagick-main/MagickCore/memory-private.h:38:38: note:
 in expansion of macro ‘CACHE_LINE_SIZE’
   38 |   __builtin_assume_aligned((address),CACHE_LINE_SIZE)
      |                                      ^~~~~~~~~~~~~~~
```

Once we get proper builds on Fedora, CentOS, Ubuntu, Debian, MacOS X, and Windows, we recommend creating an archive that includes build instructions that users can download and unpack into the ImageMagick source.  From there, they can build with either  `configure` or `cmake`.  Once that is reviewed/tested/patched, we can consider supporting both builds in the ImageMagick source distribution with an eye towards removing  the `autoconf` / `automake` build.

Note a proper build includes support for building ImageMagick statically, shared, with module support, with OpenMP, with OpenCL, with PerlMagick, and with the ImageMagick unit tests.  Once all these steps are complete, the ImageMagick developers would be in a better position to adopt and maintain the `cmake` scripts.

**Comment by Cyriuz** [comment]
Alright, I'll update on my progress here as I get it to build on all platforms, however I doubt I will be able to fix everything so hopefully someone else wants to help out.

**Comment by Cyriuz** [comment]
I think I have got it working pretty solid now on most platforms and made a PR to conan to use the CMake scripts in the conan-center package as I thought that was a great way to get it used by the community before it can be accepted into ImageMagick itself.

Some things are obviously missing still like the unit tests and PerlMagick so it is still in need of contributions if someone feels like helping out in the future :)

**Reply by mmomtchev** [reply]
@Cyriuz for CMake to be usable for any build on Windows - and fully self-contained builds on Linux - all of the delegates must also have compatible CMake build systems? Do you have an idea if and how many of them are currently usable?

**Reply by Cyriuz** [reply]
How do you mean for it to be usable? The delegates are as any other dependency, expected to be built before you build ImageMagick. I personally use conan to build the dependencies so I don't know how many of them are natively CMake, but I know that most of them are working cross platform through their conan recipes. The main reason for wanting to use CMake for ImageMagick is because the current build scripts does not support a single build flow on all platforms.

**Reply by mmomtchev** [reply]
Ok, you use conan, I will check it. What I need is exactly the single build flow on all platforms. Thank you.

**Reply by Cyriuz** [reply]
I made [this](https://github.com/conan-io/conan-center-index/pull/10003) PR to conan but it was never accepted, I should try to update it now when conan 2 has been out for a while.

**Reply by mmomtchev** [reply]
@urban-warrior your current Windows build includes a script that pulls all the dependencies and adds custom VS solutions, how do you see this if/when you switch to CMake? Would you consider a solution based on conan?

**Reply by mmomtchev** [reply]
Or, eventually, the alternative is to extend the current CMake build with the capability to checkout the repositories of the projects missing from the system - if you want to avoid depending on conan?
I am willing to sink in some time in this project to bring it to completion/merging

**Reply by Cyriuz** [reply]
In my opinion the build scripts should probably not pull any dependencies. For making releases the dependencies would be installed on the build machines, and that could be done with conan.

**Reply by Cyriuz** [reply]
The only thing I do to make it work with the conan pipeline is injecting these 2 lines: https://github.com/conan-io/conan-center-index/pull/10003/files#diff-c4e974fad2fb6ef83dfbee4091d485e1a2952158d8c71f2fa2931fc3dd938b74R203 and that can probably be done by injecting those vars in the environment instead. The conanbuildinfo.cmake file that is generated by conan contains all the module information for CMake to find the dependencies. If you haven't worked with conan before feel free to ask if you need help :)

**Reply by satmandu** [reply]
I'm not familiar with conan, but can the cmake modifications be used standalone to build imagemagick with cmake?

**Reply by Cyriuz** [reply]
Yeah for sure, you just need to make sure `find_package` finds the delegates that you have installed / built. Usually any delegates installed are found automatically but if not their paths need to be added to `CMAKE_MODULE_PATH`. That's basically the only difference, that conan automates that process for you.

**Comment by mmomtchev** [comment]
The CMake portable build is here and waiting for your suggestions: https://github.com/ImageMagick/ImageMagick/pull/7232

If anyone wants to contribute, he is more than welcome.

Special kudos to @MarcoMartins86 and @Cyriuz who made this possible.

As this is a very sizable PR, and especially given my particular situation, it might not be merged soon, but I am willing to continue to maintain it. I am the author/maintainer of the `magickwand.js` and I will be its first user.

**Reply by mmomtchev** [reply]
@wbehrens-on-gh @vincent-nv

**Comment by mmomtchev** [comment]
The build instructions can be found here:

https://github.com/mmomtchev/ImageMagick/blob/cmake-portable-build/README.CMake.md

```
git clone -b cmake-portable-build  https://github.com/mmomtchev/ImageMagick.git
```

**Reply by mmomtchev** [reply]
I intend continue to maintain this build to date with the latest version.

[![CMake portable build](https://github.com/mmomtchev/ImageMagick/actions/workflows/cmake-build.yml/badge.svg?branch=cmake-portable-build)](https://github.com/mmomtchev/ImageMagick/actions/workflows/cmake-build.yml)

**Comment by mmomtchev** [comment]
I invite everyone who is interested in the CMake build to try it. It has huge advantages for cross-platform projects that include ImageMagick as a library. It comes with a built-in `conanfile` and can use both system-installed libraries and conan-provided libraries and has been tested on Linux, macOS, Windows and WASM.

What is currently missing:
* Dynamic modules via `libltdl`
* and a number of exotic libraries such as autotrace, FLIF, Flashpix and `libumem` - which are not available on conan and cannot be easily tested in Github Actions

If you are using any of these features, you are welcome to contribute the missing part to the CMakeFiles - for everything besides the dynamic modules it is as simple as identifying the CMake targets and/or the eventual `pkg-config` name - and finding a way to test this support this in Github Actions.

Also, it should be decided, before this is set in stone - because other packages will depend on it - is how to name the CMake targets and especially how to name the targets given the compilation options. For example something like `ImageMagick::MagickCore` for any `MagickCore` and `ImageMagick::MagickCore-HDR-16` for applications that expect this configuration. Currently only the generic targets are implemented.
